keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,740 @@
+create a `PersistentVolumeClaim` or `PVC` from a previously taken VolumeSnapshot.
+This is done by pointing the `Spec.dataSource` parameter of the `PVC` to an existing
+`VolumeSnapshot` instance. There is no logic that validates whether the original
+volume mode of the `PVC`, whose snapshot was taken, matches the volume mode of 
+the newly created `PVC` that is being created from the existing `VolumeSnapshot`.
+This KEP proposes a solution to prevent unauthorized conversion of the volume
+mode during such an operation. 
+
+## Motivation
+
+Malicious users may expose a vulnerability in the kernel by exploiting
+this gap. 
+Here is an example of how a malicious user can exploit this gap to crash the
+kernel.
+1. User creates a `PVC` with `volumeMode: Block` and runs a pod with it.
+2. User writes malformed ext4 data to it (simple dd)
+3. User takes snapshot of this volume.
+4. User creates a `PVC` with `volumeMode: Filesystem` from the above snapshot.
+5. User uses this `PVC` in a pod.
+    1. kubelet tries to mount it during pod creation. If there is a CVE in the
+       kernel, the user can crash it.
+
+Note that, as of this writing, there is no known CVE in the kernel that a
+malicious user can exploit. However CVE's are regularly discovered that affect
+filesystems. For example https://access.redhat.com/security/cve/cve-2020-12655
+allows an attacker to trigger a DoS attack on the kernel.
+This proposal aims to prevent a security vulnerability in the event that a CVE 
+is discovered.
+
+We cannot simply block this operation as some backup vendors try to create a 
+volume with the exact same mode as the original volume but may need to do the
+conversion for efficiency.
+An example workflow of a backup vendor could look like:
+1. Assume the original `PVC` is created with `volumeMode: Filesystem`.
+2. During backup, the backup software will create a `PVC` from a `VolumeSnapshot`
+with `volumeMode: Block`. 
+This steps needs volume mode conversion. The purpose of creating this `PVC` with 
+block mode is to be able to copy data efficiently and save it to a backup target.
+3. The `PVC` created in the previous step is temporary and will be deleted after
+data is copied.
+4. Finally at restore time, another `PVC` will be created with `volumeMode: Filesystem`.
+
+
+### Goals
+
+Define a mechanism to mitigate the vulnerability of restoring volumes without
+hampering valid use cases.
+
+### Non-Goals
+
+Design that is generic and can be extended to other storage related security
+aspects.
+
+## Proposal
+
+The proposal aims to mitigate this issue by modifying the `VolumeSnapshotContent`
+API spec as well as the control flows of `snapshot-controller` and `external-provisioner`.
+`VolumeSnapshotContent` API will include a field that denotes the volume mode of
+the volume that the snapshot was created from. 
+This proposal also introduces a new annotation on the `VolumeSnapshotContent` resource
```

Comment:
if we own the API, why make this an annotation?

Reply:
We did think about adding this to API as well, but thought an annotation is simpler. I'm open to adding this to API. So we can add `allowVolumeModeChange` to `VolumeSnapshotClass`. This field will be copied to `allowVolumeModeChange` field in `VolumeSnapshotContent` by the snapshot-controller.

@jsafrane what do you think?